src/agent/src/rpc.rs

```diff
@@ -2016,6 +2055,140 @@ mod tests {
         assert!(result.is_err(), "expected add arp neighbors to fail");
     }
 
+    #[tokio::test]
+    async fn test_do_write_stream() {
+        #[derive(Debug)]
+        struct TestData<'a> {
+            create_container: bool,
+            has_fd: bool,
+            has_ptm: bool,
+            break_pipe: bool,
+
+            container_id: &'a str,
+            exec_id: &'a str,
+            data: Vec<u8>,
+            result: Result<protocols::agent::WriteStreamResponse>,
+        }
+
+        impl Default for TestData<'_> {
+            fn default() -> Self {
+                TestData {
+                    create_container: true,
+                    has_fd: true,
+                    has_ptm: true,
+                    break_pipe: false,
+
+                    container_id: "1",
+                    exec_id: "2",
+                    data: vec![1, 2, 3],
+                    result: Ok(WriteStreamResponse {
+                        len: 3,
+                        ..WriteStreamResponse::default()
+                    }),
+                }
+            }
+        }
+
+        let tests = &[
+            TestData::default(),
+            TestData {
+                create_container: false,
+                result: Err(anyhow!(crate::sandbox::ERR_INVALID_CONTAINER_ID)),
+                ..TestData::default()
+            },
+            TestData {
+                container_id: "8181",
+                result: Err(anyhow!(crate::sandbox::ERR_INVALID_CONTAINER_ID)),
+                ..TestData::default()
+            },
+            TestData {
+                data: vec![],
+                result: Ok(WriteStreamResponse {
+                    len: 0,
+                    ..WriteStreamResponse::default()
+                }),
+                ..TestData::default()
+            },
+            TestData {
+                break_pipe: true,
+                result: Err(anyhow!("Broken pipe (os error 32)")),
```

Comment:
One point I'd like feedback on is whether we should be asserting error messages that come from the operating system. It's something that might have the potential to change outside of the control of this repository.

Reply:
I am not that comfortable doing this, but if we rely on that for something else already, then we better ensure a change on the message won't also break the "something else" relying on this.

Reply:
I agree, I'm not too comfortable with it either. I've removed it for now. I'm open to ideas for how to test the error case of [this line](https://github.com/bradenrayhorn/kata-containers/blob/test-do-write-stream/src/agent/src/rpc.rs#L588). The ideas I've got seem like they'd complicate the test more than the value it'd provide.

Reply:
The best solution is to convert the `std::io::Error` to an O/S error (numeric `errno`) value in the test. You could do something like this in the test code (no error checking below but just shows the concept):

```rust
let result: std::io::Error = some_function();

// Convert the Error to a numeric value
let errno_value = e.raw_os_error().unwrap();

assert_eq!(errno_value, libc::EPIPE, "{}", msg);
```

Reply:
Thanks for the info @jodh-intel, I've updated the test to make its assertions in a similar method.